### DISCUSSION TITLE
Colab speed and user experience improvement

### DISCUSSION THREAD

**Discussion by ackvf** [TheLastBen/fast-stable-diffusion#2018] [Show and tell]
Hey guys, I forked your notebook and improved it. It was confusing for me and I needed a way to better manage my models, so I tweaked it a bit, but I don't understand python and SD at all, so I was wondering if you could take a look, give me some feedback, and perhaps imported some of the tweaks back into your notebook too?

The main difference is that I improved some comments and documentation (such as documenting ControlNets), streamlined some flows (xFormers setting) and split the notebook into two parts:

- A1111 notebook
https://colab.research.google.com/drive/15y0-QZM32atrnO05GVOGVP7ryPZQrl1x?usp=sharing

- Companion model downloader
https://colab.research.google.com/drive/1eZbKtuP4XbC6wOfxmu_2_UtjLmbq0Z6a?usp=sharing

Having it in a separate file makes it more flexible as you can continue running your A1111 notebook while downloading and experimenting with models and extensions.

This also allows to place several model downloads into a queue to be executed one after another, since the companion colab provides multitude of execution cells (meaning I can add another model download, while one download is already in progress).

**Comment by i-huzaifa-arshad** [comment]
Nice stuff. Glad someone came with better approach than mine

**Reply by wordbrew** [reply]
To be honest, Fast A1111 has become incredibly slow and buggy for me lately. Not sure what's going on, but as a Colab pro member, it hangs so often in the Start Stable Diffusion cell that I'm considering moving away from this repo entirely. Hangs even on brand new installs. It's gotten a bit unusable for me and I have no idea what caused it.

**Reply by TheLastBen** [reply]
don't blame the repo, blame colab, starting sd is slow because of installed extensions, gdrive speed limitation and slow colab CPU, if you want faster SD, use my Runpod notebooks or my Paperspace notebooks, find them in the main page of the repo

**Reply by ackvf** [reply]
Interesting, I have never experienced hangs during startup. I wonder what the cause might be. Using Colab Pro, but also free on another account.

**Reply by ackvf** [reply]
@TheLastBen I wonder if it would help to first copy the gdrive to colab workspace disk, then launch the app. Same for models. It's especially noticeable when changing models. Chaching them to colab disk would be much faster.

**Reply by TheLastBen** [reply]
@ackvf copying from gdrive to the workspace would take a lot of time

**Comment by TheLastBen** [comment]
it's great work, but for unexperienced users it seems a bit complicated, I like to keep it as simple as possible to avoid complications.

**Reply by ackvf** [reply]
That's for sure. How this works is still a mystery to me. But honestly, I think that I actually made it simpler in some sense (at least those parts that I struggled with), but you could say that I also made it more complex in some other sense - that's the unfortunate side effect of being the author. In my eyes, it's simple 🤣

BTW I further improved the startup time by skipping unnecessary steps.

One thing that was very missing was a way to turn off xFormers. This is how I started this all in the first place, and then I needed a better way to download many models directly to gdrive.

**Reply by TheLastBen** [reply]
You can disable xformers simply by removing --xformers in the last cell, and at the moment it is removed

**Reply by ackvf** [reply]
Yes, but I added a checkbox for that. Doing that in the command itself is not really for an unexperienced user, and I was unexperienced at that moment, hence why I added it as an option 😁

**Reply by TheLastBen** [reply]
If a user wants to turn off xformers that means they are a bit experienced and know what they're doing, I take care of the necessary arguments and implements them in a way the webui functions the best.

**Reply by i-huzaifa-arshad** [reply]
One thing I love the most is reverse engineering your code. It's a great way of learning. I want to run InvokeAi webui on colab. Just wanted to use their outpainting just for once. Hopefully I will manage to make a colab to run it. Or you can do this @TheLastBen ?

**Reply by ackvf** [reply]
@TheLastBen That is not true actually. When I started, I was upset and confused that running the same prompt will not yield same result. I searched and at one point someone suggested to check my xFormers setting, which I had no idea what they were talking about, so I spent more hours figuring that out, finally finding that in the command line of your notebook, which led into more hours to learn how to do python and colab notebooks.

**Reply by ackvf** [reply]
Ideally I would like to control xFormers from the webui itself. IIUC it allows to generate images faster.

My idea is that during experimenting, I would enable xFormers and once I settle on a preferred composition and design, I will disable xFormers and fine tune the prompt and settings. - does that sound legit or is my understanding wrong?

**Reply by ackvf** [reply]
Honestly, civitai is full of confused people asking why they were not able to reproduce the art on display, where they are usually followed by a suggestion to turn off "xFormers". I don't think at all that this is a common knowledge, but it's such a common issue that it should be brought to the attention right from the UI or the Colab - which will then incentive people to search for the meaning of it.

**Comment by anonderpling** [comment]
@ackvf 
> you can continue running your A1111 notebook while downloading and experimenting with models and extensions.

You mean like [batchlinks webui extension](https://github.com/etherealxx/batchlinks-webui) does? Your solution requires uploading files to drive from a separate colab, meaning your other colab instance won't have immediate local access to them; if you use batchlinks they _should_ be available immediately from the cache, and it downloads the model previews as well.

I use batchlinks mainly for SD models (and occasionally for loras), as they're the largest files; smaller files also get downloaded directly in the same colab with the unofficial Civitai helper extension ([Civitai](https://civitai.com/models/16768/civitai-helper-sd-webui-civitai-extension), [github](https://github.com/butaixianran/Stable-Diffusion-Webui-Civitai-Helper)); the advantage here is that while the Civitai helper doesn't download as fast as the batchlinks extension does, it downloads other data that's useful as well (for example, it grabs the prompts for the preview images). After downloading files, I tell the civitai helper to fetch metadata for those files as well, so I have the prompts for the previews. After this, I also download all the previews to my personal server, and set up a redirect extension so the previews don't time out my connection[^2][^3]. 

Alternatively, just use a download manager[^1] for your drive.

[^1]: I made [this](https://colab.research.google.com/gist/anonderpling/6b859dc8e2c6e3055a266414a4d90b3d/aria2c-webui.ipynb) ([source](https://gist.github.com/anonderpling/6b859dc8e2c6e3055a266414a4d90b3d)) using the same aria2 used by batchlinks webui extension, and [ziahamza/webui-aria2](https://github.com/ziahamza/webui-aria2) for pasting links in easier; the default path is the A1111 models/Stable-diffusion directory. You will have to choose the directory the files get downloaded in for other types, as I mainly used it for models. To make that simpler, you could create symlinks at `/models` or some such to use shorter paths.

[^2]: I use [Redirector](https://addons.mozilla.org/en-US/firefox/addon/redirector) firefox extension (choice of extension mainly because it works for firefox mobile) with [these settings](https://gist.github.com/anonderpling/a5a8bfd1557a88a28db076b0afc806f6). You'll have to update the domain name in that config. If you don't have a server of your own, you'll need a secondary tunnel (the whole point of this tunnel is to avoid slowing down your main A1111 tunnel); if you have an SSH client installed, I recommend using [localhost.run](https://localhost.run/), as they don't require registration and it needs nothing else. Alternatively, cloudflare tunnel or localtunnel will also work.

[^3]: It's a bit more complicated than this, since I have it automated. I use git lfs from huggingface to keep a copy of your models directory, and you can do a sparse checkout *without* the actual LFS tracked files, resulting in a small-ish ~300mb directory of just the previews, which I then recompress with imagemagick; the resulting directory is what I actually redirect to, however my server has very little bandwidth available, so I'm reluctant to make it public...
